crates/bevy_pbr/src/render/mesh_preprocess.wgsl

```diff
@@ -43,20 +74,103 @@ struct PreprocessWorkItem {
+#endif
+
+#ifdef FRUSTUM_CULLING
+// Data needed to cull the meshes.
+//
+// At the moment, this consists only of AABBs.
+@group(0) @binding(5) var<storage> mesh_culling_data: array<MeshCullingData>;
+
+// The view data, including the view matrix.
+@group(0) @binding(6) var<uniform> view: View;
+
+// Returns true if the view frustum intersects a sphere, specified in world
+// space.
+fn view_frustum_intersects_sphere(sphere_center_3: vec3<f32>, sphere_radius: f32) -> bool {
+    let sphere_center_4 = vec4(sphere_center_3, 1.0);
+    for (var i = 0; i < 5; i += 1) {
+        if (!maths::sphere_is_inside_frustum_plane(
+            view.frustum[i],
+            sphere_center_4,
+            sphere_radius
+        )) {
+            return false;
+        }
+    }
+    return true;
+}
+
+// Returns true if the view frustum intersects an oriented bounding box (OBB).
+//
+// `aabb_center.w` should be 1.0.
+fn view_frustum_intersects_obb(
+    model: mat4x4<f32>,
+    aabb_center: vec4<f32>,
+    aabb_half_extents: vec3<f32>,
+) -> bool {
+
+    for (var i = 0; i < 5; i += 1) {
+        // Calculate relative radius of the sphere associated with this plane.
+        let plane_normal = view.frustum[i];
+        let relative_radius = dot(
+            abs(
+                vec3(
+                    dot(plane_normal, model[0]),
+                    dot(plane_normal, model[1]),
+                    dot(plane_normal, model[2]),
+                )
+            ),
+            aabb_half_extents
+        );
+
+        // Check the frustum plane.
+        if (!maths::sphere_is_inside_frustum_plane(plane_normal, aabb_center, relative_radius)) {
+            return false;
+        }
+    }
+
+    return true;
+}
+#endif
+
 @compute
 @workgroup_size(64)
 fn main(@builtin(global_invocation_id) global_invocation_id: vec3<u32>) {
+    // Figure out our instance index. If this thread doesn't correspond to any
+    // index, bail.
     let instance_index = global_invocation_id.x;
     if (instance_index >= arrayLength(&work_items)) {
         return;
     }
 
     // Unpack.
-    let mesh_index = work_items[instance_index].input_index;
+    let input_index = work_items[instance_index].input_index;
     let output_index = work_items[instance_index].output_index;
-    let model_affine_transpose = current_input[mesh_index].model;
+    let model_affine_transpose = current_input[input_index].model;
     let model = maths::affine3_to_square(model_affine_transpose);
 
+    // Cull if necessary.
+#ifdef FRUSTUM_CULLING
+    let aabb_center = mesh_culling_data[input_index].aabb_center.xyz;
+    let aabb_half_extents = mesh_culling_data[input_index].aabb_half_extents.xyz;
+
+    // Do a quick sphere-based frustum cull.
+    let sphere_center = model * vec4(aabb_center, 1.0);
+    let sphere_radius = length(maths::mat4x4_to_mat3x3(model) * aabb_half_extents);
+    if (!view_frustum_intersects_sphere(sphere_center.xyz, sphere_radius)) {
+        return;
+    }
```

Comment:
For some reason I wonder if this is helpful on the GPU. It could be that all code has to be executed anyway, unless all waves go this way?

Reply:
Oh, interesting, you're right, I bet this can be skipped.

Reply:
I skip that check now.